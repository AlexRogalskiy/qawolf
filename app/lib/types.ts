// GraphQL Types
export type AuthenticatedUser = {
  access_token: string;
  user: User;
};

export type DeploymentEnvironment = "all" | "preview" | "production";

export type Env = { [name: string]: string };

export type Environment = {
  id: string;
  name: string;
};

export type EnvironmentVariable = {
  created_at: string;
  environment_id: string;
  id: string;
  name: string;
  value: string;
};

export type Integration = {
  github_repo_name: string;
  id: string;
  slack_channel: string | null;
  team_name: string | null;
  type: "github" | "slack";
};

export type Invite = {
  email: string;
  id: string;
  team_id: string;
  wolf_name: string;
  wolf_number: number;
  wolf_variant: string;
};

export type RunStatus = "created" | "fail" | "pass";

export type Run = {
  code: string;
  completed_at: string | null;
  created_at: string;
  current_line: number | null;
  environment_id: string | null;
  id: string;
  logs_url: string | null;
  started_at: string | null;
  status: RunStatus;
  suite_id: string | null;
  test_id: string;
  video_url: string | null;
};

export type Runner = {
  api_key: string | null;
  ws_url: string | null;
};

export type ShortSuite = {
  created_at: string;
  environment_id: string | null;
  environment_variables: string | null;
  id: string;
  team_id: string;
  trigger_color: string | null;
  trigger_id: string | null;
  trigger_name: string | null;
};

export type Suite = ShortSuite & {
  runs: SuiteRun[];
};

export type SuiteRun = {
  completed_at: string | null;
  gif_url: string | null;
  id: string;
  is_test_deleted: boolean;
  started_at: string | null;
  status: RunStatus;
  test_id: string;
  test_name: string;
};

export type ShortTeam = {
  id: string;
  name: string;
};

export type Team = ShortTeam & {
  alert_integration_id: string | null;
  api_key: string;
  helpers: string;
  inbox: string;
  is_email_alert_enabled: boolean;
  is_enabled: boolean;
  next_trigger_id: string;
  renewed_at: string | null;
};

export type ShortTest = {
  id: string;
  name: string;
};

export type Test = ShortTest & {
  code: string;
  deleted_at: string | null;
  is_enabled: boolean;
  updated_at: string;
  version: number;
};

export type TestHistoryRun = {
  created_at: string;
  id: string;
  started_at: string;
  status: RunStatus;
};

export type TestSummaryRun = {
  created_at: string;
  gif_url: string | null;
  id: string;
  status: RunStatus;
};

export type TestSummary = {
  gif_url: string | null;
  last_runs: TestSummaryRun[];
  test_id: string;
};

export type Trigger = {
  color: string;
  created_at: string;
  deployment_branches: string | null;
  deployment_environment: DeploymentEnvironment | null;
  deployment_integration_id: string | null;
  environment_id: string | null;
  id: string;
  name: string;
  next_at: string | null;
  repeat_minutes: number | null;
};

export type User = {
  avatar_url: string | null;
  email: string;
  id: string;
  onboarded_at: string | null;
  teams: ShortTeam[];
  wolf_name: string;
  wolf_number: number;
  wolf_variant: string;
};

// Component Types
export type AuthMode = "logIn" | "signUp";

export type CreateCode = {
  code: string;
  test_id: string | null;
};

export type Modal =
  | "createTest"
  | "environments"
  | "deleteTests"
  | "teamSettings"
  | "triggers";

export type NavigationOption = "code" | "logs" | "helpers";

export type NavigationType = "dark" | "light";

export type SelectedTest = {
  id: string;
  name: string;
};

export type Side = "left" | "right";

export type TestTriggers = {
  test_id: string;
  trigger_ids: string[];
};

export type TriggerFields = {
  deployment_branches: string | null;
  deployment_environment: DeploymentEnvironment | null;
  deployment_integration_id: string | null;
  environment_id: string | null;
  name: string;
  repeat_minutes: number | null;
};

export type ValueProp = {
  detail: string;
  message: string;
};

export type Wolf = {
  name: string;
  number: number;
  variant: string;
};

// Runner Types
export type CodeUpdate = {
  code: string;
  generated?: boolean;
  test_id: string;
  version: number;
};

export type Log = {
  message: string;
  severity: "info" | "warning" | "error";
  timestamp: string;
};

export type RunOptions = {
  code: string;
  end_line?: number;
  env: Env;
  helpers?: string;
  restart: boolean;
  run_id?: string;
  start_line?: number;
  test_id?: string;
  version: number;
};

export type RunProgress = {
  code: string;
  completed_at: string | null;
  current_line: number | null;
  run_id?: string;
  start_line?: number;
  status: RunStatus;
};

// State Types
type SignUp = {
  inviteId?: string | null;
  redirectUri?: string | null;
};

type ModalState = {
  name: Modal | null;
  teamId?: string;
  testIds?: string[];
  tests?: SelectedTest[];
};

export type State = {
  dashboardUri: string | null;
  editorSidebarWidth: number;
  email: string | null;
  environmentId: string | null;
  modal: ModalState;
  run: RunOptions | null;
  signUp: SignUp;
  teamId: string | null;
  toast: {
    error?: boolean;
    expiresIn?: number;
    message: string;
  } | null;
};
