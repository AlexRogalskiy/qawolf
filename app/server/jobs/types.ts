export const JOB_TYPES = [
  "checkPending",
  "deleteOldEmails",
  "deleteRunners",
  "deployRunners",
  "orchestrateRunners",
  "orchestrateTriggers",
  "restartRunners",
] as const;

export type Job = typeof JOB_TYPES[number];
