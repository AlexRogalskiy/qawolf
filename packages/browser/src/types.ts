declare module "puppeteer/lib/USKeyboardLayout" {
  export interface KeyDefinition {
    code: string;
    key: string;
    keyCode?: number;
    location?: number;
    shiftKey?: string;
    shiftKeyCode?: number;
  }

  export interface KeyDefinitionsType {
    [key: string]: KeyDefinition;
  }

  const map: KeyDefinitionsType;
  export default map;
}
