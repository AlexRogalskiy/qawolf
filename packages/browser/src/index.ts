export { Page as PuppeteerPage } from "puppeteer";
export * from "./browser";
export * from "./find";
export * from "./retry";
export * from "./strokes";
