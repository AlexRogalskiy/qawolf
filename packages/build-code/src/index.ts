export * from "./buildInitialCode";
export { stepToSelector } from "./buildSelector";
export * from "./buildVirtualCode";
export * from "./CodeReconciler";
