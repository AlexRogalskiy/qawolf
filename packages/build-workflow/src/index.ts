// export * from "./buildClickSteps";
// export * from "./buildScrollSteps";
// export * from "./buildSelectSteps";
// export * from "./buildSteps";
// export * from "./buildTypeSteps";
// export * from "./buildWorkflow";
