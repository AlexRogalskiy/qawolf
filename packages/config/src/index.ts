import dotenv from "dotenv";

const dotEnvPath = process.env.DOTENV_PATH;
dotenv.config(dotEnvPath ? { path: dotEnvPath } : {});

const testUrl = process.env.TEST_URL || "http://localhost:5000/";

const parseBool = (value: string | undefined) => {
  return !!value && value !== "false";
};

const parseNumber = (value: string | undefined) => {
  if (typeof value === "undefined") return value;

  return parseFloat(value);
};

export const CONFIG = {
  chromeExecutablePath: process.env.CHROME_EXECUTABLE_PATH,
  dataAttribute: process.env.DATA_ATTRIBUTE || null,
  headless: parseBool(process.env.HEADLESS),
  keepBrowserOpen: parseBool(process.env.KEEP_BROWSER_OPEN),
  logPath: process.env.LOG_PATH,
  saveBrowserWsPath: process.env.SAVE_BROWSER_WS_PATH,
  screenshotPath: process.env.SCREENSHOT_PATH,
  sleepAfterEach: parseNumber(process.env.SLEEP_AFTER_EACH),
  testUrl,
  useLocalModule: parseBool(process.env.USE_LOCAL_MODULE)
};
