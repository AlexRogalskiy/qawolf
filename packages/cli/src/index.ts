#!/usr/bin/env node

import { buildTest } from "@qawolf/build-test";
// import { buildWorkflow } from "@qawolf/build-workflow";
import { Browser } from "@qawolf/browser";
import { CONFIG } from "@qawolf/config";
import { logger } from "@qawolf/logger";
import { Runner } from "@qawolf/runner";
import { Workflow } from "@qawolf/types";
import program from "commander";
import { outputFile, outputJson, readJson } from "fs-extra";
import { snakeCase } from "lodash";
import { resolve } from "path";
import { runTest } from "./runTest";
import { parseUrl, getUrlRoot } from "./utils";

// program
//   .command("build <eventsPath> <name>")
//   .description("build a test from events")
//   .action(async (eventsPath, name) => {
//     const sourcePath = resolve(eventsPath);
//     logger.verbose(`read events from ${sourcePath}`);
//     const events = await readJson(sourcePath);

//     const destPath = `${process.cwd()}/.qawolf`;
//     const formattedName = snakeCase(name);
//     const destWorkflowPath = `${destPath}/workflows/${formattedName}.json`;
//     const destTestPath = `${destPath}/tests/${formattedName}.test.js`;

//     logger.verbose(`build workflow -> ${destTestPath}`);
//     const workflow = buildWorkflow(events, formattedName);
//     await outputJson(destWorkflowPath, workflow, { spaces: " " });

//     logger.verbose(`build test -> ${destTestPath}`);
//     const test = buildTest(workflow);
//     await outputFile(destTestPath, test, "utf8");

//     process.exit(0);
//   });

let recordCommand = program
  .command("record <url> [name]")
  .description("record a workflow and create a test");

if (CONFIG.development) {
  recordCommand = recordCommand.option(
    "-e, --events",
    "record events (for debugging)"
  );
}

recordCommand.action(async (urlArgument, optionalName, cmd) => {
  const url = parseUrl(urlArgument);
  const name = optionalName || getUrlRoot(url);
  logger.verbose(`record url "${url.href}"`);

  const browser = await Browser.create({ record: true, url: url.href });
  await browser.waitForClose();

  const destPath = `${process.cwd()}/.qawolf`;
  const formattedName = snakeCase(name);

  if (cmd.events) {
    const destEventsPath = `${destPath}/events/${formattedName}.json`;
    logger.verbose(`create events "${name}" -> ${destEventsPath}`);
    await outputJson(destEventsPath, browser.events, { spaces: " " });
  } else {
    logger.verbose(`create test "${name}"`);
  }

  process.exit(0);
});

program
  .command("run [name]")
  .description("run a workflow")
  .action(async name => {
    const workflowPath = `${process.cwd()}/.qawolf/workflows/${snakeCase(
      name
    )}.json`;
    const workflow = (await readJson(workflowPath)) as Workflow;
    const runner = await Runner.create(workflow);

    await runner.run();
    await runner.close();

    process.exit(0);
  });

program
  .command("test [name]")
  .description("run a test")
  .action(async name => {
    const results = await runTest(name ? snakeCase(name) : null);
    const success = results.numFailedTestSuites < 1;

    process.exit(success ? 0 : 1);
  });

program.allowUnknownOption(false);

program.parse(process.argv);

if (!process.argv.slice(2).length) {
  console.log("\n");
  program.outputHelp();
}
