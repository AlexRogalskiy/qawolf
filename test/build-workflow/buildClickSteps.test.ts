import { pick } from 'lodash';
import { loadFixtures } from '../loadFixtures';
import { buildClickSteps } from '../../src/build-workflow';
import { ElementEvent } from '../../src/types';

describe('buildClickSteps', () => {
  let radioEvents: ElementEvent[];
  let loginEvents: ElementEvent[];
  let selectEvents: ElementEvent[];

  beforeAll(async () => {
    radioEvents = (await loadFixtures('radioInputs')).events;
    loginEvents = (await loadFixtures('login')).events;
    selectEvents = (await loadFixtures('selectNative')).events;
  });

  it('builds one click per group of mousedown/click events', () => {
    const steps = buildClickSteps(loginEvents);
    expect(
      // snapshot to test it includes the original event index
      steps.map(step => pick(step, 'htmlSelector', 'index')),
    ).toMatchSnapshot();
  });

  it('skips click triggered by Enter', () => {
    const steps = buildClickSteps(loginEvents);

    expect(steps.length).toEqual(3);

    // click on login link
    expect(steps[0].target.attrs.href).toEqual('/login');

    // click on initial input
    expect(steps[1].target.attrs.id).toEqual('username');

    // click logout
    expect(steps[2].target.attrs.qaw_innertext).toEqual('Log out');
  });

  it('skips click on select', () => {
    const steps = buildClickSteps(selectEvents);
    expect(steps.length).toBe(0);
  });

  it('prefers clicks on inputs', () => {
    const steps = buildClickSteps(radioEvents);
    expect(steps.map(step => step.cssSelector)).toMatchSnapshot();
  });
});
