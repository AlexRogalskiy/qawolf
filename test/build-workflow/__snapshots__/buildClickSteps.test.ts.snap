// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildClickSteps builds one click per group of mousedown/click events 1`] = `
Array [
  Object {
    "htmlSelector": "<ul qaw_innertext=\\"Buttons Checkbox inputs Content editables Date pickers Images Infinite scroll Large Log in Radio inputs Selects Text inputs Time pickers\\"><li qaw_innertext=\\"Log in\\"><a href=\\"/login\\" qaw_innertext=\\"Log in\\"></a></li></ul>",
    "index": 0,
  },
  Object {
    "htmlSelector": "<div id=\\"root\\" qaw_innertext=\\"To log in, use tomsmith as the username, and SuperSecretPassword! as the password. Username Password Log in\\"><form qaw_innertext=\\"Username Password Log in\\"><input autocomplete=\\"off\\" id=\\"username\\" type=\\"text\\" value=\\"\\" qaw_labels=\\"Username\\" /></form></div>",
    "index": 2,
  },
  Object {
    "htmlSelector": "<div id=\\"root\\" qaw_innertext=\\"Secure Area Log out\\"><button qaw_innertext=\\"Log out\\"><p qaw_innertext=\\"Log out\\"></p></button></div>",
    "index": 30,
  },
]
`;

exports[`buildClickSteps prefers clicks on inputs 1`] = `
Array [
  "[data-qa='html-radio']",
  "[data-qa='html-radio-group'] [value='dog']",
  "[data-qa='material-radio'] [value='yes']",
  "[data-qa='material-radio-group'] [value='green']",
]
`;
